### PR TITLE
!!! FEATURE: `NodeTypeManager` remove NodeType fallback handling

### PR DESCRIPTION
**Pull request by nezaniel**
This moves node type fallback handling from the CR to the outer applications, e.g. Neos.

!!! Nodes now have an **optional node type** which is null if the NodeTypeName is not known to the NodeTypeManager.

Neos has been adjusted to
* ask the NodeTypeManager for the node type
* deal with NodeTypeNotFound itself by applying fallback to Neos.Neos:Fallback

For 3rd party applications relying on fallback NodeTypes, there now is the `Neos\Neos\Utility\NodeTypeWithFallbackProvider trait` available

Related: #4447

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php [hunk 1]
```diff
@@ -57,13 +57,17 @@ public function __construct(
...
     public function mapNodeRowToNode(
         array $nodeRow,
         DimensionSpacePoint $dimensionSpacePoint,
         VisibilityConstraints $visibilityConstraints
     ): Node {
+        try {
+            $nodeType = $this->nodeTypeManager->getNodeType($nodeRow['nodetypename']);
+        } catch (NodeTypeNotFoundException $exception) {
+            $nodeType = null;
+        }
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php]: Hmm im not sure i like to use exceptions as control flow here (exceptions are always expensive - one could argue that this is an edgecase and not the path taken in 90% of the cases ... so well fine as well)

But how about

A: adding a new method wich is nullable? (Or make `getNodeType` nullable -> no i think rather not?)

```suggestion
        $nodeType = $this->nodeTypeManager->findNodeType($nodeRow['nodetypename']);
```

B: ask if the nt mngr "has" the node:

```suggestion
        $nodeType = $this->nodeTypeManager->hasNodeType($nodeRow['nodetypename'])
            ? $this->nodeTypeManager->getNodeType($nodeRow['nodetypename'])
            : null;
```

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php [hunk 2]
```diff
@@ -50,15 +50,22 @@ public function __construct(
...
+        public NodeAggregateClassification $classification,
+        /**
+         * The node's node type name; always set, even if unknown to the NodeTypeManager
+         */
+        public NodeTypeName $nodeTypeName,
+        /**
+         * The node's node type, null if unknown to the NodeTypeManager
+         * @deprecated Ask the NodeTypeManager instead
+         */
+        public ?NodeType $nodeType,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php]: +1 as we discussed its now nullable and deprecated!

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use Neos\Flow\Annotations as Flow;
+use Neos\Neos\Domain\Service\NodeTypeNameFactory;
+
+/**
+ * Utility trait for retrieving node types for nodes with a built-in fallback mechanism
+ */
+trait NodeTypeWithFallbackProvider
+{
+    #[Flow\Inject]
+    protected ContentRepositoryRegistry $contentRepositoryRegistry;
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: traits with flow inject annotation - hmmm something rings a bell @fcool said to me that `inject` methods here work more reliable. Should we also rename the property to something prefixed only for this trait like `$NodeTypeWithFallbackProvider_contentRepositoryRegistry`?

i did notice that currently you are removing the dependency from the classes where the trait is also injected

```diff
-    #[Flow\Inject]
-     protected ContentRepositoryRegistry $contentRepositoryRegistry;
```

i just dislike such intransitive dependencies ^^ or are my concerns are unfounded?

but good idea to have this trait <3

File: Neos.ContentRepository.StructureAdjustment/src/Adjustment/LoadNodeTypeTrait.php
Comment (mhsdesign) [inline Neos.ContentRepository.StructureAdjustment/src/Adjustment/LoadNodeTypeTrait.php:1]: haha - this was the exact reason i wanted to defeat the fallback nt. I guess the `LoadNodeTypeTrait` can now be removed (ill can take care of it)

There are only 5 usages.

File: Neos.ContentRepositoryRegistry/Classes/Factory/NodeTypeManager/DefaultNodeTypeManagerFactory.php [hunk 4]
```diff
@@ -24,8 +24,7 @@ function () {
                 $configuration = $this->configurationManager->getConfiguration('NodeTypes');
                 return $this->nodeTypeEnrichmentService->enrichNodeTypeLabelsConfiguration($configuration);
             },
-            $this->nodeLabelGeneratorFactory,
-            $options['fallbackNodeTypeName'] ?? null,
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Factory/NodeTypeManager/DefaultNodeTypeManagerFactory.php:28]: i guess we have to also remove the yaml config for it?

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    protected function getNodeType(Node $node): NodeType
+    {
+        $nodeTypeManager = $this->contentRepositoryRegistry->get($node->subgraphIdentity->contentRepositoryId)->getNodeTypeManager();
+
+        try {
+            return $nodeTypeManager->getNodeType($node->nodeTypeName);
+        } catch (NodeTypeNotFoundException) {
+            return $nodeTypeManager->getNodeType(NodeTypeNameFactory::forFallback());
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: the fallback nt `Neos.Neos:FallbackNode` will (currently) be not configureable anymore - but im totally fine with it.

File: Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php [hunk 6]
```diff
@@ -116,53 +124,55 @@ protected function buildView()
     protected function setUp(): void
     {
         parent::setUp();
+        $this->setUpCRRegistry();
+        $contentRepositoryId = ContentRepositoryId::fromString('default');
+        $this->iInitializeContentRepository($contentRepositoryId->value);
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php]:
```suggestion
        $contentRepositoryId = ContentRepositoryId::fromString('default');
```

imo we dont need the behat cr initialization stuff here?

**Review by mhsdesign — COMMENTED** [review]
THANK YOU SO MUCH for this quick implementation.

I did a full visual review - it was a little harder because you also adjusted more code to use the `NodeTypeNameFactory` and avoid now magic strings like `'Neos.Neos:Document'` but thank you for that as well.

I have a few points to discuss left - but overall i approve mostly already ;)

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php (hunk 3 above)
Comment (fcool) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: you remember correctly - but this is only valid for the annotations.
Trouble with annotations is, that their namespace gets lost during proxy building. But attributes should be always fine.

Comment (mhsdesign) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: ah okay ;) and what about my concern of that this property pollutes now the class? I think it should only be visible to the trait (even though that is not possible - unless we obscurity the name ^^)

the trait is meant to be used in userland code... and i guess if another trait is using the same property name but without type and legacy annotation we have a problem?

File: Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php [hunk 7]
```diff
@@ -148,7 +148,8 @@ private function createVariantRecursivelyInternal(int $level, NodeAggregateId $p
 
         foreach ($childNodes as $childNode) {
             if ($childNode->classification->isRegular()) {
-                if ($childNode->nodeType->isOfType('Neos.Neos:Document')) {
+                $childNodeType = $contentRepository->getNodeTypeManager()->getNodeType($childNode->nodeTypeName);
+                if ($childNodeType->isOfType('Neos.Neos:Document')) {
```
Comment (dlubitz) [inline Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php:192]: I'm wondering why do we need the `Neos.Neos:Document` here? Isn't it this a Neos thing? As `Neos.ContentRepositoryRegistry` has no dependency to Neos we can't use the const of NodeTypeNameFactory unfortunately.

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php [hunk 8]
```diff
@@ -34,13 +34,13 @@
  *
  * @api Note: The constructor is not part of the public API
  */
-final class Node
+final readonly class Node
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:37]: I remember very loosely, that @bwaidelich had some issues with readonly classes and proxy builder. 🤷‍♂️ But I could be totally wrong.

**Review by dlubitz — APPROVED** [review]
Looks good to me by reading! Thank you

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php (hunk 8 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:37]: @robertlemke luckily fixed these issues now ^^ (see testcases in flow :D)

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php]: We should adjust the deprecation notice in that we probably want to replace the $nodeType object and not remove the property: eg the concept of this big NodeType god object directly on the node is deprecated, but there will always be some kind of schema information - even if less powerful - on the node

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php]: @nezaniel agrees ^^

File: Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php (hunk 7 above)
Comment (nezaniel) [reply Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php:192]: Agreed, the Registry should not care about document nodes. Would you open a new issue for this, please?

File: Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php (hunk 6 above)
Comment (nezaniel) [reply Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php]: Well as the Helper needs a Registry, we should make use of the new capabilities. Perhaps an alternative less coupled to Behat would be nice...

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php (hunk 5 above)
Comment (nezaniel) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: Did anyone ever use that? We could make it configurable here, if needed

File: Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php]: ah you mean the nodeHelper is used to ask the cr of the nodeTypeManager, to get the nodeType?

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: yes i agree - IF we want this. We should assure that we have cleaned up any leftover yaml artifacts

File: Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php (hunk 6 above)
Comment (nezaniel) [reply Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php]: exactly, no direct calls to Node::nodetype anymore

File: Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php (hunk 7 above)
Comment (dlubitz) [reply Neos.ContentRepositoryRegistry/Classes/Command/ContentCommandController.php:192]: See: https://github.com/neos/neos-development-collection/issues/4488

File: Neos.ContentRepositoryRegistry/Classes/Utility/ContentRepositoryRegistryProvider.php [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Neos\ContentRepositoryRegistry\Utility;
+
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * Utility trait for retrieving node types for nodes with a built-in fallback mechanism
+ */
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Utility/ContentRepositoryRegistryProvider.php]:
```suggestion
```

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php]: has been adjusted where available

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php (hunk 3 above)
Comment (nezaniel) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: the registry injection has been extracted to a separate trait to clean this up

File: Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php (hunk 5 above)
Comment (nezaniel) [reply Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php]: I cleaned up the artifacts :D

File: Neos.ContentRepositoryRegistry/Classes/Factory/NodeTypeManager/DefaultNodeTypeManagerFactory.php (hunk 4 above)
Comment (nezaniel) [reply Neos.ContentRepositoryRegistry/Classes/Factory/NodeTypeManager/DefaultNodeTypeManagerFactory.php:28]: done

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php]: documentation has been adjusted

File: Neos.ContentRepositoryRegistry/Classes/Utility/ContentRepositoryRegistryProvider.php (hunk 9 above)
Comment (nezaniel) [reply Neos.ContentRepositoryRegistry/Classes/Utility/ContentRepositoryRegistryProvider.php]: Docblock has been adjusted

**Review by mhsdesign — CHANGES_REQUESTED** [review]
Thanks for adjusting to the feedback from the weekly.

But i think we misunderstood each other about the traits.
With https://github.com/neos/neos-development-collection/pull/4466/commits/364a230f4e2568f9a7f59a9aad18c17927fc2f91 you added the `ContentRepositoryRegistryProvider` and i think you wanted to implement bastians suggestion, but this would require an abstract method `getContentRepositoryById`.

Instead i propose to delay this discussion - or do nothing fancy here at all ~and just import the `ContentRepositoryRegistry` directly in the `NodeTypeWithFallbackProvider` as prefixed variable.~

~This pattern - while not especially pretty - has already been used across the codebase and we might just as well do it here too before blocking this pr any longer with this discussion.~

Edit: i just found out that phpstan will perfectly analyze the traits, so we only have to use the property in the trait and declare it at another place: https://phpstan.org/blog/how-phpstan-analyses-traits

I implemented the suggestion here - feel free to merge it into this pr: :D
https://github.com/neos/neos-development-collection/pull/4559

File: Neos.ContentRepository.StructureAdjustment/src/Adjustment/LoadNodeTypeTrait.php
Comment (mhsdesign) [reply Neos.ContentRepository.StructureAdjustment/src/Adjustment/LoadNodeTypeTrait.php:1]: adjusted in https://github.com/neos/neos-development-collection/pull/4559

**Comment by mhsdesign** [comment]
The neos ui has to be adjusted as well. I removed the Headline nodeType from the demo site to test the fallback but the whole request fails (in the neos backend):

```
Call to a member function isOfType() on null

Exception Code | 0
-- | --
Error
Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Neos_Neos_Ui_ContentRepository_Service_WorkspaceService.php
210
Packages/Application/Neos.Neos.Ui/Classes/ContentRepository/Service/WorkspaceService.php
```

This seems to be tested nowhere as the Neos Ui e2e tests pass and the neos ui has no phpstan 😢 


Also the fusion rendering in the frontend is not really successful after removing the Headline

image

Im not sure, but this is not the behavior of Neos 8.x -> the node should just be not rendered.

thats because we access the nodeType in fusion:

the content case needs to be adjusted to (or do we actually need a better way of getting the nodes nodetype in fusion?)
```
prototype(Neos.Neos:ContentCase) < prototype(Neos.Fusion:Case) {
  default {
    @position = 'end'
    condition = true
    type = ${node.nodeType.name || "Neos.Neos:FallbackNode"}
  }
}
```

**Comment by mhsdesign** [comment]
I think we’ll need a NodeType for node helper for eel aswell? `${Neos.Node.getNodeType(node)}`

im not sure how we handle `node.nodeType` is the use deprecated? Should we migrate Fusion already? And for how long is it guaranteed that the node knows it’s nodeType if fetched via the official ways - until Neos 10 and the well remove it?

Those questions need to be answered so that our users now what to expect and we know when to deprecate/remove it.

File: .composer.json [hunk 10]
```diff
@@ -14,7 +14,7 @@
   "scripts": {
     "lint:phpcs-psr12": "../../bin/phpcs --colors --standard=PSR12 ./Neos.ContentGraph.DoctrineDbalAdapter/src ./Neos.ContentGraph.PostgreSQLAdapter/src ./Neos.ContentRepository.BehavioralTests/Classes ./Neos.ContentRepository.TestSuite/Classes ./Neos.ContentRepository.Core/Classes ./Neos.Neos/Classes",
     "lint:phpcs": [
-      "@lint:phpcs-psr12 --exclude=Generic.Files.LineLength"
+      "@lint:phpcs-psr12 --exclude=Generic.Files.LineLength,PSR1.Files.SideEffects"
```
Comment (bwaidelich) [inline .composer.json:18]: Why do we need to change the linter rules for this change? that makes me suspicious :)

**Review by bwaidelich — COMMENTED** [review]
Makes a lot of sense to me, I just wonder about the changed linter rules

File: .composer.json (hunk 10 above)
Comment (mhsdesign) [reply .composer.json:18]: This was a problem with phpcs and `readonly` - but it should have been fixed with: https://github.com/squizlabs/PHP_CodeSniffer/issues/3727

Comment (mhsdesign) [reply .composer.json:18]: but they are stuck a bit in their release workflow. We have to wait patiently ;)

Excluding `PSR1.Files.SideEffects` is actually no big deal as we write proper php code generally ^^

```
FILE: /Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1 | WARNING | A file should declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it should execute logic with side effects, but should not do both. The first symbol is defined on line 37 and the first side effect is on line 37.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Review by mhsdesign — APPROVED** [review]
In combination with https://github.com/neos/neos-ui/pull/3632 the neos ui e2e tests succeed locally

**Pull request merged**